### PR TITLE
Financial Chart: CandleStickRenderer, HighLowRenderer, Basic Financial API

### PR DESCRIPTION
**Pull request by raven2cz**
Pull request according to issue #294. Screenshots attached in the issue. The samples and unit tests too.

### CODE REVIEW

**Review by RalphSteinhagen — CHANGES_REQUESTED** [review]
Very nice pull-request and only minor comments that should be addressed, mainly:

a) to open/keep the candle-stick and hi-lo renderer also compatible with other more generic DataSet implementation (e.g. MultiDimDoubleDataSet or any other DataSet with nDim>=6)

b) you may consider adding some JavaDoc and '@author' and references to the classes as documentation, acknowledgement of your work, and so that everyone knows long-term who has produced these nice renderer additions.

In any case, very nice PR and if you agree to the minor changes mentioned above, I'd approve and sponsor this PR and long-term maintenance within chart-fx.

File: chartfx-chart/src/test/java/de/gsi/chart/renderer/spi/financial/CandleStickRendererTest.java
Comment (wirew0rm) [inline chartfx-chart/src/test/java/de/gsi/chart/renderer/spi/financial/CandleStickRendererTest.java:87]: This test appends to the paintAfterEPS list from the junit thread and also from the JavaApplicationThread in the setup routine. Also as per setup(), the paintAfterEPS list should not be empty, but it seems it is at the beginning of this method?

**Review by wirew0rm — CHANGES_REQUESTED** [review]
Hey there,

first of all also from me a big thank you for your contribution, this all looks very nice. It looks like there's only a few small things left before this can be merged.

I think Ralph already pointed out the bigger API ones, so mine are mostly smaller nits and suggestions. Please feel free to disagree with any of them, the only thing that should really be changed before mege is the number of dimensions for the OhlcvDataSet (in addition to Ralphs points).

I also like the samples, if you want you can also add some nice screenshots into doc/pics and add them to the gallery in the Readme.md, otherwise we will do that after the merge is done.

File: chartfx-chart/src/test/java/de/gsi/chart/utils/StyleParserTest.java
Comment (raven2cz) [reply chartfx-chart/src/test/java/de/gsi/chart/utils/StyleParserTest.java:80]: corrected.

File: chartfx-dataset/src/main/java/de/gsi/dataset/spi/financial/OhlcvDataSet.java
Comment (raven2cz) [reply chartfx-dataset/src/main/java/de/gsi/dataset/spi/financial/OhlcvDataSet.java:24]: After the refactoring any common dataset can be used with both renderers, e.g.MultiDimDoubleDataSet. IOhlcv and IOhlcvItem are necessary to provide chart-fx API to the middle layer of any financial frameworks. The ohlcv filled data structures are more complex and hold large memory which cannot be copy to additional data blocks of chart-fx datasets or series. Better way is to provide interface and hide complex implementation by implement middle layer between complex financial frameworks and chart libraries.
In an actual implementation, there is opened a possibility to use any dataset; like MultiDimDoubleDataSet or others. But not for all features! If developer wants to additional advantages like addons and fast lookup to ohlcv items, the dataset has to implement additional new interfaces _AttributeModelAware_ and _IOhlcvItemAware_.  Basic usage of MultiDimDoubleDataSet is not recommended for Extension points and PaintBars. In any cases, the common datasets are supported now.

File: chartfx-samples/pom.xml
Comment (raven2cz) [reply chartfx-samples/pom.xml:55]: Dependency removed. Use from chartfx-math.

File: chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/css/FinancialColorSchemeConstants.java
Comment (raven2cz) [reply chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/css/FinancialColorSchemeConstants.java:17]: renamed.

File: chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/CandleStickRenderer.java
Comment (raven2cz) [reply chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/CandleStickRenderer.java:155]: Used iMin and iMax.

File: chartfx-chart/src/main/java/de/gsi/chart/Chart.java
Comment (raven2cz) [reply chartfx-chart/src/main/java/de/gsi/chart/Chart.java:847]: Still kept! I looked to the chart impl., that there exists some a proprietary solution for css styling of it. But it needs much more work in chart for correct functionality. This financial commit cannot change chart structure much more. It needs create some separate issue for it; to improve styling of chart title. In actual case, I have to keep this setter for both dark themes; the title is not visible!

File: chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/css/FinancialColorSchemeConfig.java
Comment (raven2cz) [reply chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/css/FinancialColorSchemeConfig.java:19]: Yes, styling is long run, I know. In addition, this is not just fixed styling for renderer. It has to be styling per dataset, it is reason why it is dynamically defined per dataset not as css styling of renderer itself. 
In addition, FinancialColorSchemeConfig changes and configure chart domain object. 
I expect some progress with it in next versions of financial components. This is just small beginning. Your help with DESIGN will be welcome here. In actual state of implementation, I kept this solution with config class.

File: chartfx-chart/src/test/java/de/gsi/chart/plugins/ScreenshotTest.java
Comment (raven2cz) [reply chartfx-chart/src/test/java/de/gsi/chart/plugins/ScreenshotTest.java:140]: ChartTest is added. It was missing. So, chart class has test now. There are just new methods added only with simple asserts.

File: chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/HighLowRenderer.java
Comment (raven2cz) [reply chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/HighLowRenderer.java:35]: moved.

File: chartfx-samples/src/main/java/de/gsi/chart/samples/RunChartSamples.java
Comment (raven2cz) [reply chartfx-samples/src/main/java/de/gsi/chart/samples/RunChartSamples.java:88]: Prefix added.

File: chartfx-samples/src/main/java/de/gsi/chart/samples/financial/ColorChooserSample.java
Comment (raven2cz) [reply chartfx-samples/src/main/java/de/gsi/chart/samples/financial/ColorChooserSample.java:19]: deleted.

File: chartfx-samples/src/main/java/de/gsi/chart/samples/financial/AbstractBasicFinancialApplication.java
Comment (raven2cz) [reply chartfx-samples/src/main/java/de/gsi/chart/samples/financial/AbstractBasicFinancialApplication.java:215]: abstract now.

File: chartfx-chart/src/test/java/de/gsi/chart/renderer/spi/financial/CandleStickRendererTest.java
Comment (raven2cz) [reply chartfx-chart/src/test/java/de/gsi/chart/renderer/spi/financial/CandleStickRendererTest.java:87]: removed, already tested inside previous test.

**Comment by raven2cz** [comment]
The review impl. changes are pushed now. I will see how about restyle again....after few next commits.....
Pictures are added too. Some javadoc too. Additional features will be part of next PR.

**Comment by raven2cz** [comment]
Done

File: chartfx-chart/src/main/java/de/gsi/chart/Chart.java
Comment (RalphSteinhagen) [reply chartfx-chart/src/main/java/de/gsi/chart/Chart.java:847]: OK... we will tackle this during one of the next PRs related to CSS theming of the charts which we intended to do anyway. Please bear with us when some of the direct setter/getters will change to CSS properties though. :-)

**Comment by RalphSteinhagen** [comment]
@raven2cz two quick questions before merging:

In the renderer you wrote:

```Java
if (ds.getDimension() > 7)
                continue;
```

shouldn't be this 

```Java
if (ds.getDimension() < 7)
                continue;
```

ie. that any DataSet that has less than 7 dimensions is rejected? The way '> 7 '  would reject only DataSets with `dim = 8, 9, 10, ...`.

Second questions: do you want the commit structure preserved or can we squash-and-merge your PR?

Thanks again for your patience and for the follow-up commits. I feel this is a very nice additions and would be happy to merge if you are OK with this as well.

File: chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/CandleStickRenderer.java
Comment (wirew0rm) [inline chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/CandleStickRenderer.java:131]:
```suggestion
            if (ds.getDimension() < 7) // dataset has to have at least 7 dimensions, time -> (open, high, low, close, volume, oi)
```

Additional dimensions should be ignored, but if you have less than 7 you will get index out of bounds errors later on. As the renderer only seems to use the first 4  y values, `< 5 `could also be a valid choice.

Probably the same for HiLowRenderer?

File: chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/HighLowRenderer.java
Comment (wirew0rm) [inline chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/financial/HighLowRenderer.java:126]: see CandlestickRenderer.

**Review by wirew0rm — CHANGES_REQUESTED** [review]
Changes looks good, thanks also for the feedback on the more open points and the screenshots.

I think that one compare operation should work the other way round, but then we're good to go.

EDIT: seems like I somehow missed Ralph's comment :/

**Review by RalphSteinhagen — DISMISSED** [review]

**Review by wirew0rm — DISMISSED** [review]

**Review by wirew0rm — APPROVED** [review]
Very nice, lets get this merged :rocket:

**Review by RalphSteinhagen — APPROVED** [review]
LGTM!

**Pull request merged**
